pylzy/lzy/env/base.py

```diff
@@ -0,0 +1,54 @@
+from __future__ import annotations
+
+from abc import ABC
+from dataclasses import is_dataclass, fields
+from typing import Dict, Any, TypeVar, Union
+from typing_extensions import Self, TypeGuard
+
+
+class NotSpecifiedType:
+    """
+    >>> NotSpecified = NotSpecifiedType()
+    >>> NotSpecified or 1
+    1
+    >>> 1 and NotSpecified
+    NotSpecified
+    """
+    def __repr__(self):
+        return 'NotSpecified'
+
+    def __bool__(self):
+        return False
+
+
+T = TypeVar('T')
+NotSpecified = NotSpecifiedType()
+EnvironmentField = Union[NotSpecifiedType, T]
+
+
+def is_specified(val: Union[NotSpecifiedType, T]) -> TypeGuard[T]:
+    return val is not NotSpecified
+
+
+class Deconstructible(ABC):
```

Comment:
Why do we even need this interface?
We have decided that we will not support merge of envs.

Reply:
Well, it giving us a simple way to copy every instance of our envs.
And simply make non-modifying  `env.with_fields(...)`.
I'm using this at few places.
Also, look, I'm using it at LzyEnvironment and Provisioning.

Reply:
imho, it's over-engineering